File: benches/bench_serialize_and_deserialize.py
import numpy as np
import msgpack
import msgpack_numpy as m
import time

raw_unpackb = msgpack.unpackb
m.patch()


def bench_serialize(value, iterations):
    start_time = time.perf_counter()
    for _ in range(iterations):
        msgpack.packb(value)
    total_time = time.perf_counter() - start_time

    return total_time / iterations * 1000


def bench_deserialize_view(value, iterations):
    packed = msgpack.packb(value)

    start_time = time.perf_counter()
    for _ in range(iterations):
        msgpack.unpackb(packed)
    total_time = time.perf_counter() - start_time

    return total_time / iterations * 1000


def bench_deserialize_owned(value, iterations):
    packed = msgpack.packb(value)

    start_time = time.perf_counter()
    for _ in range(iterations):
        [array.copy() for array in msgpack.unpackb(packed)]
    total_time = time.perf_counter() - start_time

    return total_time / iterations * 1000


def bench_raw_unpack(value, iterations):
    packed = msgpack.packb(value)

    start_time = time.perf_counter()
    for _ in range(iterations):
        raw_unpackb(packed)
    total_time = time.perf_counter() - start_time

    return total_time / iterations * 1000


def bench_array_construction(value, iterations):
    packed = msgpack.packb(value)
    decoded = raw_unpackb(packed)

    start_time = time.perf_counter()
    for _ in range(iterations):
        [m.decode(item) for item in decoded]
    total_time = time.perf_counter() - start_time

    return total_time / iterations * 1000


def benchmark_case(dtype, array_size, array_count, iterations):
    array = np.arange(array_size, dtype=dtype)
    arrays = [array] * array_count
    print(
        f"{np.dtype(dtype).name},{array_size},{array_count},"
        f"{bench_serialize(arrays, iterations):.3f},"
        f"{bench_deserialize_view(arrays, iterations):.3f},"
        f"{bench_deserialize_owned(arrays, iterations):.3f},"
        f"{bench_raw_unpack(arrays, iterations):.3f},"
        f"{bench_array_construction(arrays, iterations):.3f}"
    )


def benchmark_large_array(dtype, iterations):
    array = np.zeros((1024, 2048), dtype=dtype)
    arrays = [array] * 10
    print(
        f"{np.dtype(dtype).name},"
        f"{bench_serialize(arrays, iterations):.3f},"
        f"{bench_deserialize_view(arrays, iterations):.3f},"
        f"{bench_deserialize_owned(arrays, iterations):.3f},"
        f"{bench_raw_unpack(arrays, iterations):.3f},"
        f"{bench_array_construction(arrays, iterations):.3f}"
    )


if __name__ == "__main__":
    iterations = 10

    print(
        "dtype,array_size,array_count,serialize_ms,deserialize_view_ms,"
        "deserialize_owned_ms,raw_unpack_ms,array_construction_ms"
    )
    for dtype in (np.float32, np.float16):
        benchmark_case(dtype, 1000, 10000, iterations)
        benchmark_case(dtype, 100, 100000, iterations)

    print("large-array benchmark: shape=(1024, 2048), array_count=10")
    print(
        "dtype,serialize_ms,deserialize_view_ms,deserialize_owned_ms,"
        "raw_unpack_ms,array_construction_ms"
    )
    benchmark_large_array(np.float32, iterations)
    benchmark_large_array(np.float16, iterations)

File: benches/bench_serialize_and_deserialize.rs
use half::f16;
use msgpack_numpy::{CowNDArray, NDArray};
use ndarray::{Array1, Array2};
use serde::Serialize;
use std::hint::black_box;
use std::time::{Duration, Instant};

fn bench_serialize<T: Serialize>(value: &T, iterations: u32) -> Duration {
    let start = Instant::now();
    for _ in 0..iterations {
        black_box(rmp_serde::to_vec_named(value).unwrap());
    }
    start.elapsed() / iterations
}

fn bench_ndarray_deserialize(value: &[NDArray], iterations: u32) -> Duration {
    let buf = rmp_serde::to_vec_named(value).unwrap();

    let start = Instant::now();
    for _ in 0..iterations {
        black_box(rmp_serde::from_slice::<Vec<NDArray>>(&buf).unwrap());
    }
    start.elapsed() / iterations
}

fn bench_cow_deserialize(value: &[NDArray], iterations: u32) -> Duration {
    let buf = rmp_serde::to_vec_named(value).unwrap();

    let start = Instant::now();
    for _ in 0..iterations {
        black_box(rmp_serde::from_slice::<Vec<CowNDArray<'_>>>(&buf).unwrap());
    }
    start.elapsed() / iterations
}

fn milliseconds(duration: Duration) -> f64 {
    duration.as_secs_f64() * 1_000.0
}

fn benchmark_case(
    dtype: &str,
    array_size: usize,
    array_count: usize,
    array: NDArray,
    iterations: u32,
) {
    let arrays = vec![array; array_count];
    println!(
        "{dtype},{array_size},{array_count},{:.3},{:.3}",
        milliseconds(bench_serialize(&arrays, iterations)),
        milliseconds(bench_ndarray_deserialize(&arrays, iterations)),
    );
}

fn benchmark_zero_copy(dtype: &str, array: NDArray, iterations: u32) {
    let arrays = vec![array; 10];
    let buf = rmp_serde::to_vec_named(&arrays).unwrap();
    let cow_arrays: Vec<CowNDArray<'_>> = rmp_serde::from_slice(&buf).unwrap();
    let borrowed_arrays = cow_arrays
        .iter()
        .filter(|array| match array {
            CowNDArray::F16(array) => array.is_view(),
            CowNDArray::F32(array) => array.is_view(),
            _ => false,
        })
        .count();

    println!(
        "{dtype},{:.3},{:.3},{:.3},{borrowed_arrays}/{}",
        milliseconds(bench_serialize(&arrays, iterations)),
        milliseconds(bench_ndarray_deserialize(&arrays, iterations)),
        milliseconds(bench_cow_deserialize(&arrays, iterations)),
        arrays.len(),
    );
}

fn main() {
    let iterations = 10;

    println!("dtype,array_size,array_count,serialize_ms,ndarray_deserialize_ms");
    benchmark_case(
        "f32",
        1000,
        10000,
        NDArray::F32(Array1::range(0., 1000., 1.).into_dyn()),
        iterations,
    );
    benchmark_case(
        "f32",
        100,
        100000,
        NDArray::F32(Array1::range(0., 100., 1.).into_dyn()),
        iterations,
    );
    benchmark_case(
        "f16",
        1000,
        10000,
        NDArray::F16(Array1::range(0., 1000., 1.).mapv(f16::from_f32).into_dyn()),
        iterations,
    );
    benchmark_case(
        "f16",
        100,
        100000,
        NDArray::F16(Array1::range(0., 100., 1.).mapv(f16::from_f32).into_dyn()),
        iterations,
    );

    println!("zero-copy benchmark: shape=(1024, 2048), array_count=10");
    println!("dtype,serialize_ms,ndarray_deserialize_ms,cow_deserialize_ms,cow_borrowed");
    benchmark_zero_copy(
        "f32",
        NDArray::F32(Array2::zeros((1024, 2048)).into_dyn()),
        iterations,
    );
    benchmark_zero_copy(
        "f16",
        NDArray::F16(Array2::from_elem((1024, 2048), f16::ZERO).into_dyn()),
        iterations,
    );
}
